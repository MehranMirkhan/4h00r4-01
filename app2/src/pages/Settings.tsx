import React from "react";
import { connect } from "react-redux";
import { IonList, IonItem } from "@ionic/react";

import Page from "src/widgets/Page";
import SelectOne from "src/widgets/SelectOne";
import { AppState } from "src/state";
import { langSelector, setLang } from "src/state/settings";

export function Settings({ lang, setLang }: ISettings) {
  return (
    <Page title={"Settings"} showBack={true}>
      <IonList>
        <IonItem>
          <SelectOne
            label={"Language"}
            value={lang}
            options={[
              { name: "English", value: "en" },
              { name: "فارسی", value: "fa" }
            ]}
            onChange={setLang}
          />
        </IonItem>
      </IonList>
    </Page>
  );
}

interface ISettings {
  lang?: string;
  setLang?: (lang: string) => void;
}

const props = (state: AppState) => ({ lang: langSelector(state) });
const actions = { setLang };

export default connect(props, actions)(Settings);
