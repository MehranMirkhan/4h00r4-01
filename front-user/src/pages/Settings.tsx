import React from 'react';
import {
  IonPage, IonHeader, IonToolbar,
  IonButtons, IonMenuButton, IonTitle,
  IonContent, IonBackButton, IonList,
  IonItem, IonLabel, IonSelect, IonSelectOption
} from '@ionic/react';


const SettingsPage: React.FC = () => {
  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonButtons slot="start">
            <IonMenuButton />
          </IonButtons>
          <IonButtons slot="end">
            <IonBackButton defaultHref="/home" />
          </IonButtons>
          <IonTitle>Setting</IonTitle>
        </IonToolbar>
      </IonHeader>

      <IonContent>
        <SettingsItems />
      </IonContent>
    </IonPage>
  );
};

const SettingsItems = () => {
  return (
    <IonList>
      <IonItem>
        <IonLabel position="floating">Language</IonLabel>
        <IonSelect>
          <IonSelectOption value="en" selected>English</IonSelectOption>
          <IonSelectOption value="fa">فارسی</IonSelectOption>
        </IonSelect>
      </IonItem>
    </IonList>
  );
};

export default SettingsPage;