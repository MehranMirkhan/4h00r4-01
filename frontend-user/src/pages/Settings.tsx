import React from "react";
import { IonList, IonItem } from "@ionic/react";
import { useTranslation } from "react-i18next";

import Page from "src/widgets/Page";
import { SelectOne } from "src/widgets/FormItems";
import { useDispatch, useSelector } from "react-redux";
import { setLang } from "src/reducers/settings.reducer";

export default function() {
  const { t } = useTranslation();
  const settings = useSelector((state: State) => state.settings);
  const dispatch = useDispatch();
  const onLangChange = (lang: string) => {
    dispatch(setLang(lang));
    setTimeout(() => {
      window.location.reload();
    }, 500);
  };
  return (
    <Page title={t("Settings")} showBack={true}>
      <IonList>
        <IonItem>
          <SelectOne
            label={t("Language")}
            value={settings.lang}
            options={[
              { name: "English", value: "en" },
              { name: "فارسی", value: "fa" }
            ]}
            onChange={onLangChange}
          />
        </IonItem>
      </IonList>
    </Page>
  );
}
