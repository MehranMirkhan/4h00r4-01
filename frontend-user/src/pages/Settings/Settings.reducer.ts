
import Storage from '../../Storage';

export enum SettingsActions {
  SET_LANG = "settings/SET_LANG",
  RESET = "settings/RESET",
};


const initialState = {
  lang: "fa",
};

export default (state = initialState, action: any) => {
  let newState;
  switch (action.type) {
    case SettingsActions.SET_LANG:
      newState = {
        ...state,
        lang: action.payload,
      };
      Storage.setObject("settings", newState);
      break;
    case SettingsActions.RESET:
      newState = {
        ...initialState,
        ...(action.payload ? action.payload : {}),
      };
      Storage.setObject("settings", newState);
      break;
    default:
      newState = state;
  }
  return newState;
};

// Actions

export const initialLoad = () => (dispatch: any) => {
  return Storage.getObject("settings").then((v: any) => {
    if (!!v) {
      console.log("Settings loaded:", v);
      dispatch(reset(v));
    }
  });
};

export const reset = (state?: any) => ({
  type: SettingsActions.RESET,
  payload: state ? state : undefined,
});

export const setLang = (lang: String) => ({
  type: SettingsActions.SET_LANG,
  payload: lang,
});
