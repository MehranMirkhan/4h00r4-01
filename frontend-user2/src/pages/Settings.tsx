import React, { useContext } from "react";
import { IonList, IonItem } from "@ionic/react";
import { useTranslation } from "react-i18next";

import Page from "src/widgets/Page";
import { SelectOne } from "src/widgets/FormItems";
import { storageContext } from "src/providers/StorageProvider";

export default function() {
  const { t } = useTranslation();
  const {
    storageState: { settings },
    storageActions: { setSettings }
  } = useContext(storageContext);
  const onSettingChange = (key: string) => (value: string) => {
    setSettings({ ...settings, [key]: value });
    window.location.reload();
  };
  return (
    <Page title={t("Settings")} showBack={true}>
      <IonList>
        <IonItem>
          <SelectOne
            label={t("Language")}
            value={settings.lang}
            options={[
              { name: "English", value: "en" },
              { name: "فارسی", value: "fa" }
            ]}
            onChange={onSettingChange("lang")}
          />
        </IonItem>
      </IonList>
    </Page>
  );
}
