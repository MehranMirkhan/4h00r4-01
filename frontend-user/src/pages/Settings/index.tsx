import React from 'react';
import {
  IonPage, IonHeader,
  IonContent, IonList,
  IonItem, IonLabel, IonSelect, IonSelectOption
} from '@ionic/react';
import { useSelector, useDispatch } from 'react-redux';
import { SelectChangeEventDetail } from '@ionic/core';
import { Translate } from 'react-localize-redux';

import { setLang } from './Settings.reducer';
import config from '../../app.config.json';
import { translate } from '../../utils';
import Toolbar from '../../components/Toolbar';


const SettingsPage: React.FC = () => {
  return (
    <IonPage>
      <IonHeader>
        <Toolbar title="pages.settings.title"/>
      </IonHeader>

      <IonContent>
        <SettingsItems />
      </IonContent>
    </IonPage>
  );
};

const SettingsItems = () => {
  const settings = useSelector((state: any) => state.settings);
  const dispatch = useDispatch();
  const langProps = (lang: String) => ({
    value: lang,
    selected: settings.lang === lang,
  });
  const onLangChange = (event: CustomEvent<SelectChangeEventDetail>) => {
    dispatch(setLang(event.detail.value));
    window.location.reload();
  };
  return (
    <IonList>
      <IonItem>
        <IonLabel position="floating"><Translate id="pages.settings.lang" /></IonLabel>
        <IonSelect onIonChange={onLangChange}
          okText={translate(settings.lang, "ok")}
          cancelText={translate(settings.lang, "cancel")}>
          {config.languages.map(l => <IonSelectOption key={l.code} {...langProps(l.code)}>{l.name}</IonSelectOption>)}
        </IonSelect>
      </IonItem>
    </IonList>
  );
};

export default SettingsPage;
