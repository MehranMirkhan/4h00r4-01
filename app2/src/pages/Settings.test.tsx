import React from "react";
import { shallow } from "enzyme";

import { Settings } from "./Settings";

describe("<Settings />", () => {
  it("should render", () => {
    const comp = shallow(<Settings />);
    expect(comp).toBeDefined();
    comp.setProps({ lang: null });
    expect(comp).toBeDefined();
    comp.setProps({ lang: [] });
    expect(comp).toBeDefined();
    comp.setProps({ lang: "en" });
    expect(comp).toBeDefined();
  });
  it("should match snapshot", () => {
    const withLangEn = shallow(<Settings lang="en" />);
    expect(withLangEn).toMatchSnapshot();
    const withLangFa = shallow(<Settings lang="fa" />);
    expect(withLangFa).toMatchSnapshot();
  });
});
