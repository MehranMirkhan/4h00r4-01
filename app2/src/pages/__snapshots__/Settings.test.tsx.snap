// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> should match snapshot 1`] = `
<_default
  showBack={true}
  title="Settings"
>
  <ForwardRef(IonList)>
    <ForwardRef(IonItem)>
      <_default
        label="Language"
        options={
          Array [
            Object {
              "name": "English",
              "value": "en",
            },
            Object {
              "name": "فارسی",
              "value": "fa",
            },
          ]
        }
        value="en"
      />
    </ForwardRef(IonItem)>
  </ForwardRef(IonList)>
</_default>
`;

exports[`<Settings /> should match snapshot 2`] = `
<_default
  showBack={true}
  title="Settings"
>
  <ForwardRef(IonList)>
    <ForwardRef(IonItem)>
      <_default
        label="Language"
        options={
          Array [
            Object {
              "name": "English",
              "value": "en",
            },
            Object {
              "name": "فارسی",
              "value": "fa",
            },
          ]
        }
        value="fa"
      />
    </ForwardRef(IonItem)>
  </ForwardRef(IonList)>
</_default>
`;
